Runtime/Native/Base/NativeClientBase.cs

```diff
@@ -0,0 +1,116 @@
+﻿using Backtrace.Unity.Model;
+using Backtrace.Unity.Model.Breadcrumbs;
+using Backtrace.Unity.Extensions;
+using System.Threading;
+
+[assembly: System.Runtime.CompilerServices.InternalsVisibleTo("Backtrace.Unity.Tests.Runtime")]
+namespace Backtrace.Unity.Runtime.Native.Base
+{
+    internal abstract class NativeClientBase
+    {
+        internal const string AnrMessage = "ANRException: Blocked thread detected.";
+        protected const string HangType = "Hang";
+        protected const string CrashType = "Crash";
+        protected const string ErrorTypeAttribute = "error.type";
+
+        /// <summary>
+        /// Determine if ANR occurred and NativeClient should report ANR in breadcrumbs
+        /// </summary>
+        protected volatile bool LogAnr = false;
+
+        // Last Backtrace client update time 
+        protected volatile internal float LastUpdateTime;
+
+        /// <summary>
+        /// Determine if the ANR background thread should be disabled or not 
+        /// for some period of time.
+        /// This option will be used by the native client implementation
+        /// once application goes to background/foreground
+        /// </summary>
+        internal volatile bool PreventAnr = false;
+
+        /// <summary>
+        /// Determine if ANR thread should exit
+        /// </summary>
+        internal volatile bool StopAnr = false;
+
+        internal Thread AnrThread;
+
+        /// <summary>
+        /// Determine if Native client should capture native crashes
+        /// </summary>
+        protected bool CaptureNativeCrashes = false;
+
+        /// <summary>
+        /// Determine if Native client should handle ANRs
+        /// </summary>
+        protected bool HandlerANR = false;
+
+        protected readonly BacktraceConfiguration _configuration;
+        protected readonly BacktraceBreadcrumbs _breadcrumbs;
+
+        private readonly bool _shouldLogAnrsInBreadcrumbs;
+
+        private object _lockObject = new object();
+        internal NativeClientBase(BacktraceConfiguration configuration, BacktraceBreadcrumbs breadcrumbs)
+        {
+            _configuration = configuration;
+            _breadcrumbs = breadcrumbs;
+            _shouldLogAnrsInBreadcrumbs = ShouldStoreAnrBreadcrumbs();
+        }
+
+        /// <summary>
+        /// Update native client state
+        /// </summary>
+        /// <param name="time">Current game time</param>
+        public void Update(float time)
+        {
+            LastUpdateTime = time;
+            lock (_lockObject)
```

Comment:
If I understand correctly, this method is run basically every frame. Locks aren't known to be exactly cheap - also, if there is an event in which the lock is awaited to be freed, wouldn't the game freeze?

Reply:
The reason behind it is to be sure we don't log something twice. a simple boolean flag might not be enough. This method might be invoked at the same time by more than 1 frame.  async/await is also unavailable here since it requires adding an additional package and maintaining it.  I thought lock is the cheapest solution that we have available.

Reply:
At least don't wait for the lock if you know nothing would happen.

```csharp
if (LogAnr) 
{
    if (Monitor.TryLock(_lockObject)) 
    {
        try 
        {
            if (LogAnr) 
            {
                // Push report
                LogAnr = false;
            }
        }
        finally
        {
            Monitor.Exit(_lockObject);
        }
    }
}
```